source/extensions/filters/http/grpc_stats/grpc_stats_filter.cc

```diff
@@ -17,66 +19,171 @@ namespace GrpcStats {
 
 namespace {
 
+// A map from gRPC service/method name to symbolized stat names
+// for the service/method.
+//
+// The expected usage pattern is that the map is populated once,
+// and can then be queried lock-free as long as it isn't being
+// modified.
+class GrpcServiceMethodToRequestNamesMap {
+public:
+  GrpcServiceMethodToRequestNamesMap(Stats::SymbolTable& symbol_table)
+      : stat_name_pool_(symbol_table) {}
+
+  void populate(const envoy::config::core::v3::GrpcMethodList& method_list) {
+    for (const auto& service : method_list.services()) {
+      Stats::StatName stat_name_service = stat_name_pool_.add(service.name());
+
+      StringMap<Grpc::Context::RequestNames>& method_map = map_[service.name()];
+      for (const auto& method_name : service.method_names()) {
+        Stats::StatName stat_name_method = stat_name_pool_.add(method_name);
+        method_map[method_name] = Grpc::Context::RequestNames{stat_name_service, stat_name_method};
+      }
+    }
+  }
+
+  absl::optional<Grpc::Context::RequestNames>
+  lookup(const Grpc::Common::RequestNames& request_names) const {
+    auto service_it = map_.find(request_names.service_);
+    if (service_it != map_.end()) {
+      const auto& method_map = service_it->second;
+
+      auto method_it = method_map.find(request_names.method_);
+      if (method_it != method_map.end()) {
+        return method_it->second;
+      }
+    }
+
+    return {};
+  }
+
+private:
+  StringMap<StringMap<Grpc::Context::RequestNames>> map_;
+  Stats::StatNamePool stat_name_pool_;
+};
+
+struct Config {
+  Config(const envoy::extensions::filters::http::grpc_stats::v3::FilterConfig& proto_config,
+         Server::Configuration::FactoryContext& context)
+      : context_(context.grpcContext()), emit_filter_state_(proto_config.emit_filter_state()),
+        whitelist_(context.scope().symbolTable()) {
+
+    switch (proto_config.per_method_stat_specifier_case()) {
+    case envoy::extensions::filters::http::grpc_stats::v3::FilterConfig::
+        PER_METHOD_STAT_SPECIFIER_NOT_SET:
```

Comment:
@htuch here's where I am (I think) detecting the difference between false and unset.

Reply:
Yeah, I think I might be wrong, but it's pretty confusing the difference between unset and default value case set (see https://developers.google.com/protocol-buffers/docs/proto3#oneof). Can you use a `BoolValue` here? This would be more consistent with the rest of the API for defaulting behaviors.